### PR TITLE
Backtrace support, part 2: Shadow VMRuntimeLimits for each stack

### PR DESCRIPTION
**Pull request by frank-emrich**
As of PR #98, each `ContinuationObject` contains an object of type `StackLimits`. In addition, there exists such an object for the main stack, it is stored in the `VMContext` and the `StackChain::MainStack` variant at the list of currently active continuations points to it. These `StackLimits` objects contain counterparts  of the stack-related fields in `VMRuntimeLimits`, namely `stack_limit` and the various `last_wasm_*` fields.

As of PR #98, the actual contents of these `StackLimits` are unused (and not updated). This changes in this PR:

While the `VMRuntimeLimits` continue to contain the stack-related information for the currently executing stack (either the main stack or a continuation), we ensure that for stacks that are *not* currently running, their corresponding `StackLimits` object now contains accurate values about their stack limits. The doc comment on `wasmtime_continuations::StackChain` describes the exact invariants that we maintain.

To ensure that these invariants hold, we need to copy some fields between the `VMRuntimeLimits` and  `StackLimits` objects when stack switching occurs. In particular, the `tc_resume` libcall now takes an additional argument: A pointer to the `StackLimits` object of the *parent* of the continuation that is about to be resume. Note that this needs to happen in the libcall itself, in order to obtain accurate values for the `last_wasm_exit_*` values in the `VMRuntimeLimits`.

### CODE REVIEW

File: crates/cranelift/src/func_environ.rs [hunk 1]
```diff
@@ -2646,7 +2647,14 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
     ) -> WasmResult<()> {
         // If the `vmruntime_limits_ptr` variable will get used then we initialize
         // it here.
-        if self.tunables.consume_fuel || self.tunables.epoch_interruption {
+        if true || self.tunables.consume_fuel || self.tunables.epoch_interruption {
```
Comment (dhil) [inline crates/cranelift/src/func_environ.rs:2650]: Idea: we should add a field to `tunables`, e.g. `tunables.resume` which we set whenever we enter `translate_resume`.

File: crates/runtime/src/continuation.rs [hunk 2]
```diff
@@ -207,6 +208,20 @@ pub fn resume(
...
+
+        (*parent_stack_limits).stack_limit = *runtime_limits.stack_limit.get();
+        (*parent_stack_limits).last_wasm_entry_sp = *runtime_limits.last_wasm_entry_sp.get();
+        // These last two values were only just updated in the `runtime_limits`
+        // because we entered the current libcall.
+        (*parent_stack_limits).last_wasm_exit_fp = *runtime_limits.last_wasm_exit_fp.get();
+        (*parent_stack_limits).last_wasm_exit_pc = *runtime_limits.last_wasm_exit_pc.get();
+
+        *runtime_limits.stack_limit.get() = (*contobj).limits.stack_limit;
+        *runtime_limits.last_wasm_entry_sp.get() = (*contobj).limits.last_wasm_entry_sp;
```
Comment (dhil) [inline crates/runtime/src/continuation.rs:222]: Isn't this zero on the initial invocation? If so, isn't that a bug?

**Review by dhil — COMMENTED** [review]
A few quick comments.

File: crates/cranelift/src/func_environ.rs (hunk 1 above)
Comment (frank-emrich) [reply crates/cranelift/src/func_environ.rs:2650]: Unfortunately, this needs to have run before we execute `translate_resume` for the first time, so that we an use the runtime limits `Variable`.

File: crates/runtime/src/continuation.rs (hunk 2 above)
Comment (frank-emrich) [reply crates/runtime/src/continuation.rs:222]: Yes and no. For a freshly created continuation, its `limits.last_wasm_entry_sp` is indeed 0 (or any other unspecific placeholder for it not being set before). But immediately after these instructions run, we resume the continuation. Because the continuation hasn't run before, this means executing its array call trampoline, which in turn sets the `last_wasm_entry_sp` in the `VMRuntimeLimits` to the correct value.
This behaviour is documented [here](https://github.com/wasmfx/wasmfxtime/pull/99/files#diff-31ba266d671a4acbdb0c7134ea9f783bab8922bd74203d5e97b94fc7853adb96R122)

File: crates/cranelift/src/func_environ.rs (hunk 1 above)
Comment (dhil) [reply crates/cranelift/src/func_environ.rs:2650]: Set it in cranelift then? Or does it need to be set when enter function is run?

Comment (frank-emrich) [reply crates/cranelift/src/func_environ.rs:2650]: The function we are looking at (`before_translate_function`) is called to perform some setup before `func_translator::parse_function_body` starts actual translation of a particular function. In the current version of this PR, this setup must run before `translate_resume` does. So the only option would be to detect inside `parse_function_body` that there is a `resume` in the function body, and then indicate to `before_translate_function` that we want `declare_vmruntime_limits_ptr` to be executed. But `parse_function_body` isn't designed to iterate over the instructions in the function twice.

Long story short: That's all rather annoying. Given that it has no runtime overhead, I decided to leave it as is for now.

**Comment by frank-emrich** [comment]
There's currently an issue with the CI, because I originally created this PR against a base branch other than `main`. I've tested things locally and will create a fresh PR after this PR is approved so that the CI gets run before merging.

**Comment by dhil** [comment]
> There's currently an issue with the CI, because I originally created this PR against a base branch other than `main`. I've tested things locally and will create a fresh PR after this PR is approved so that the CI gets run before merging.

I think closing and opening the PR again ought to do the trick.

File: crates/cranelift/src/func_environ.rs (hunk 1 above)
Comment (dhil) [reply crates/cranelift/src/func_environ.rs:2650]: OK, thanks for clarifying. It would be good to open an issue about this "problem".

File: crates/runtime/src/continuation.rs (hunk 2 above)
Comment (dhil) [reply crates/runtime/src/continuation.rs:222]: Thanks for clarifying. I think it would be good to put a comment on line 220 with a back reference to the top comment.

**Review by dhil — APPROVED** [review]
LGTM.

File: crates/cranelift/src/func_environ.rs (hunk 1 above)
Comment (frank-emrich) [reply crates/cranelift/src/func_environ.rs:2650]: #105
